Cybernetic Implants

xxalpha opened the pull request:
Augments for your brain. These give you permanent (for now) additional functionalities/abilities, such as HUDs. Ideas for implants are welcome (planning on adding X-ray and Thermals eye implants).

Need feedback for the idea and code implementation.

---
Eye Implants:
-Medical HUD
-Security HUD
-X-ray vision
-Thermal vision

Brain Implants:
-Anti-Drop
-Anti-Stun

---------
Update 1
Added X-ray and sprites by @Dorsisdwarf 
![eyeimplants](https://cloud.githubusercontent.com/assets/4951579/6546770/9faef5f2-c5b8-11e4-9038-83df01733f66.png)

---------
Update 2
Added Thermals and made implants vulnerable to EMP. EMP will make you blind and remove any vision augments, for 5 seconds.

---------
Update 3
Added Anti-Drop implant for the brain, it will prevent item dropping when activated. Sprite by @Nienhaus 
![brainimplant](https://cloud.githubusercontent.com/assets/4951579/6563905/b1439dba-c69c-11e4-9ed8-f0d0b49e4d13.png)

---------
Update 4
Added Anti-Stun implant for the brain, it will greatly reduce stun durations.
![stuntest](https://cloud.githubusercontent.com/assets/4951579/6587722/ec03d90e-c77f-11e4-9447-808b3d720a5f.gif)

Cheridan commented:
Please provide more information on what the implants do.

also 
my vision is augmented :eyeglasses:

RemieRichards commented:
Seems to change their eye colour and give them inbuilt HUD glasses.

xxalpha commented:
Updated the OP, I hope it clarifies.

Dorsisdwarf commented:
Binary Channel Cybernetic Implant please.

Makes your eyes solid white, syndicate tech 3-4, gives you binary channel.

xxalpha commented:
These are not exclusively eye implants, dorsisdwarf, but I'll note your idea.

Fox-McCloud commented:
Would recommend giving these to robotics in addition to a full surgery tool set.

Gives them something more to do than build a few borgs and a riply every round.

xxalpha commented:
I was actually thinking about giving them exclusively to research, maybe cargo as well. I suppose one or two free implants wouldn't hurt in robotics.

hornygranny commented:
these need to be EMPable somehow

Fox-McCloud commented:
@xxalpha 

I meant more they cost resources to make in the mechfab or something like that---as in, once tech levels are high enough they'll appear under their own category and robotics can make them.

xxalpha commented:
That falls under research, then.

Alek2ander commented:
Feature request: implant rejection. chemistry needs more things to do. also a DX:HR reference.

TheVekter commented:
@Alek2ander I'd feel like that'd be mindless resource management. Love the general idea of this PR though.

Pennwick commented:
I hope sunglasses implants become a thing.

Jordie0608 commented:
Changelog

hornygranny commented:
@xxalpha temporary blindness, like the old effect on people who were using thermals

xxalpha commented:
Added emp_act(), it will make you completely blind for 5 seconds.

GunHog commented:
How do we acquire the implant items?

Alek2ander commented:
Please don't use something as powerful as SEE_INFRA for this...

xxalpha commented:
I'll just remove the vulnerability.

GunHog commented:
Are you adding a way to acquire these implants, or are they going to be admin-only?

xxalpha commented:
The ones I have made so far all belong in Research. RND will be able to produce them.

RemieRichards commented:
Why are these /organ anyway? they have no reason to be /organ.

RemieRichards commented on code/modules/mob/mob_defines.dm:
```diff
@@ -156,4 +156,6 @@
-	var/last_movement = 0 // Last world.time the mob actually moved of its own accord.
\ No newline at end of file
+	var/last_movement = 0 // Last world.time the mob actually moved of its own accord.
+	var/permanent_huds = new /list() //any HUDs in this list will be ignored by HUD removing procs
```
Don't define lists like this please.
```list_variable = list()``` is our standard

RemieRichards commented on code/modules/mob/mob_defines.dm:
```diff
@@ -156,4 +156,6 @@
\ No newline at end of file
+	var/last_movement = 0 // Last world.time the mob actually moved of its own accord.
+	var/permanent_huds = new /list() //any HUDs in this list will be ignored by HUD removing procs
+	var/permanent_sight_flags = new /list() //any sight flags in this list will be ignored by removing procs
```
See Above.

RemieRichards commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -510,10 +510,16 @@
 		if(!H.druggy)		H.see_invisible = SEE_INVISIBLE_LEVEL_TWO
 	else
-		H.sight &= ~(SEE_TURFS|SEE_MOBS|SEE_OBJS)
+		if(!locate(SEE_TURFS,H.permanent_sight_flags))
```
The Vast majority of your ```if(!locate(X,Y))``` can be written as ```if(!(X in Y))``` which is the standard for finding things in lists.

RemieRichards commented on code/modules/surgery/eye_cybernetic_implant.dm:
```diff
@@ -0,0 +1,31 @@
+	location = "eyes"
+
+/datum/surgery_step/cybernetic_implant
+	accept_any_item = 1
```
This step would probably be better off by having it's tool list contain the base type path (/obj/item/organ/cybernetic_implant/eyes) instead of simply accepting any item and doing the type check yourself.

xxalpha replied:
To differentiate what kind of implant you're going to implant, as eye implants are the not the only ones. This was my first contact with surgery though and I will go over this code again.

RemieRichards replied:
That's the point xxalpha, this SURGERY can be eyes only, but this step, if you make the changes I suggested, can be used for any and all future cybernetic_implants.

RemieRichards commented on code/modules/surgery/eye_cybernetic_implant.dm:
```diff
@@ -0,0 +1,31 @@
+
+/datum/surgery_step/cybernetic_implant/success(mob/user, mob/living/carbon/target, target_zone, obj/item/tool, datum/surgery/surgery)
+	if(tool)
+		if(!istype(tool, /obj/item/organ/cybernetic_implant/eyes))
```
See Above.

xxalpha commented:
These are organs because it makes sense. What else would they be?

RemieRichards commented:
just ```obj/item/cybernetic_implant``` it's rare we use type paths JUST for organisation, and there's no real reason other than that for them to be /organ.

xxalpha commented:
Changed everything you said. Also tried to make thermal vulnerability less snowflake, if it's still not good enough I'll remove it.

xxalpha commented:
I will add nothing else to this PR.

Dorsisdwarf commented:
The anti-drop weapon seems incredibly powerful. Perhaps change it to a high chance of not dropping your item instead (~80%)? Otherwise, the disarm intent becomes useless, dualsaberers with noslips become godlike, the works.

xxalpha commented:
An EMP will cause the implant to fail and also stun the user.

GunHog commented:
Ping! Are you still working on this wonderful PR?

xxalpha commented:
Not adding content, no. I am waiting for more reviewing and merging/closing. Also fixing merge conflicts.

GunHog commented:
Well dang. If you get the merge conflicts fixed, I will help you annoy the maintainers @RemieRichards in coderbus for a review!

xxalpha commented:
Fixed merge conflicts and HUD examine not working.

Miauw62 commented on code/modules/mob/living/carbon/carbon.dm:
```diff
@@ -188,6 +188,13 @@
+	for(var/obj/item/I in contents)
+		if(istype(I,/obj/item/cybernetic_implant/eyes))
+			var/obj/item/cybernetic_implant/eyes/EFP = I
+			return EFP.flash_protect
```
You can replace this with ```var/obj/item/cybernetic_implant/eyes/EFP = locate() in src```

Fox-McCloud commented on code/modules/research/designs.dm:
```diff
@@ -381,3 +381,67 @@ datum/design/proc/CalcReliability(var/list/temp_techs)
+	build_type = PROTOLATHE
+	materials = list("$metal" = 200, "$glass" = 200, "$silver" = 200, "$gold" = 100)
+	build_path = /obj/item/cybernetic_implant/eyes/hud/medical
+	category = list("Equipement")
```
These should all be medical designs, not equipment.

Miauw62 commented on code/modules/mob/living/carbon/human/human_helpers.dm:
```diff
@@ -88,6 +88,11 @@
+		if(istype(I,/obj/item/cybernetic_implant/eyes))
+			var/obj/item/cybernetic_implant/eyes/EFP = I
+			number += EFP.flash_protect
+			break
```
See previous comment.

Miauw62 commented on code/modules/surgery/organs/cybernetic_implants.dm:
```diff
@@ -0,0 +1,216 @@
+	active = !active
+	if(active)
+		if(owner.l_hand)
+			if(NODROP in owner.l_hand.flags)
```
flags is not a list, this should be if (owner.l_hand.flags & NODROP)

xxalpha replied:
okay but it works somehow

Miauw62 commented on code/modules/surgery/organs/cybernetic_implants.dm:
```diff
@@ -0,0 +1,216 @@
+				l_hand_ignore = 0
+
+		if(owner.r_hand)
+			if(NODROP in owner.r_hand.flags)
```
same here.

Miauw62 commented on code/modules/surgery/organs/cybernetic_implants.dm:
```diff
@@ -0,0 +1,216 @@
+		else
+			var/noodles = 0
+			noodles += !l_hand_ignore && owner.l_hand ? 1 : 0
+			noodles += !r_hand_ignore && owner.r_hand ? 2 : 0
```
noodles????

Miauw62 commented on code/modules/surgery/organs/cybernetic_implants.dm:
```diff
@@ -0,0 +1,216 @@
+			owner.l_hand.flags &= ~NODROP
+		if(!r_hand_ignore && owner.r_hand)
+			owner.r_hand.flags &= ~NODROP
+		owner << "<span class='notice'>Your hands relax...</span>"
```
You can put NODROP clothing on (so you can pick up cursed stuff and put it on, only to be unable to take it off again), so you could use this by taking gloves in your hand, using this implant, putting the gloves on and then summoning an armblade to have a "free" armblade. I suggest you use direct references to these items (with a loc == owner check)

xxalpha replied:
I don't understand the solution you suggested.

Miauw62 commented on code/modules/surgery/organs/cybernetic_implants.dm:
```diff
@@ -0,0 +1,216 @@
+		if(!l_hand_ignore && owner.l_hand)
+			owner.l_hand.flags &= ~NODROP
+		if(!r_hand_ignore && owner.r_hand)
+			owner.r_hand.flags &= ~NODROP
```
Also, you can use ^= instead of &= ~

Miauw62 commented on code/modules/surgery/organs/cybernetic_implants.dm:
```diff
@@ -0,0 +1,216 @@
+		if(!l_hand_ignore && owner.l_hand)
+			owner.l_hand.flags &= ~NODROP
+		if(!r_hand_ignore && owner.r_hand)
+			owner.r_hand.flags &= ~NODROP
```
Would probably be fun if it could randomly switch this to on or something.

Miauw62 commented:
Apart from that, this seems OK.

Fox-McCloud commented:
the EMP act for the anti_drop implant should probably FORCE the person to drop whatever is in their hands; just having it turn the NODROP off isn't that big of a deal--implant needs a downside to it.

Still think these implants should be in the Mech fabricator so roboticists can perform the surgeries for these.

xxalpha commented:
@Fox-McCloud the brain implant emp_act() already stuns the user, all items will be dropped if they have NODROP.

And I'll add them to the fabricator.

Fox-McCloud commented:
@xxalpha 

Oh, sorry, missed that! Fair enough then, that'll definitely accomplish the same thing.

xxalpha commented:
Changed designs category to 'Medical Designs'.
Added designs to the mecha fabricator.
Changed Anti-Drop implant to cause the user to throw the items he is holding in random directions.

Fox-McCloud commented:
looks great =3

palpatine213 commented:
@xxalpha been playtesting your changes, and I've noticed a problem with the brain implants, mainly that you saw the person's skull open, leaving them with a ton of brute damage.
The other surgeries which use that either augment your head, which heals all that damage, or remove your brain, which then kills you. Seems like something you might want to change.

Miauw62 commented on code/modules/mob/living/carbon/life.dm:
```diff
@@ -552,10 +552,14 @@
+		if(!(SEE_MOBS in permanent_sight_flags))
+			sight &= ~SEE_MOBS
+		if(!(SEE_OBJS in permanent_sight_flags))
+			sight &= ~SEE_OBJS
```
pls use ^=

xxalpha replied:
Preposterous!

Miauw62 commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -510,10 +510,16 @@
+		if(!(SEE_MOBS in H.permanent_sight_flags))
+			H.sight &= ~SEE_MOBS
+		if(!(SEE_OBJS in H.permanent_sight_flags))
+			H.sight &= ~SEE_OBJS
```
same here

Miauw62 commented on code/modules/surgery/organs/cybernetic_implants.dm:
```diff
@@ -0,0 +1,236 @@
+	owner << "<span class='warning'>Static obfuscates your vision!</span>"
+	spawn(50)
+	owner.sight |= save_sight
+	owner.disabilities &= ~BLIND
```
^=

Miauw62 commented on html/changelogs/xxalpha-cyberimplants.yml:
```diff
@@ -0,0 +1,6 @@
+delete-after: True
+
+changes:
+  - rscadd: "Added Cybernetic Implants."
```
This could probably do with a slightly larger changelog

Miauw62 commented on code/modules/surgery/organs/cybernetic_implants.dm:
```diff
@@ -0,0 +1,236 @@
+		if(!l_hand_ignore && owner.l_hand && owner.l_hand == l_hand_obj)
+			owner.l_hand.flags ^= NODROP
+		if(!r_hand_ignore && owner.r_hand && owner.r_hand == r_hand_obj)
+			owner.r_hand.flags ^= NODROP
```
This could still cause people to have NODROP jumpsuits that they are wearing. I suggest doing this:
```
if(!r_hand_ignore && r_hand_obj in owner.contents)
    r_hand_obj.flags ^= NODROP
```
it's not perfect, but it's the best we can do.

(also make this a separate release_item() proc so you dont have to copypaste it in the emp_act())

xxalpha commented:
Tested it, should work properly now.

Lumipharon commented:
Some of these sound awfully overpowered, considering the scarcity of emps.

palpatine213 commented:
@xxalpha Turns out the cautery doesn't fix the brute you get from sawing your head open, so brain implants are extremely annoying to implant, since you have to rush to medbay right afterwards. Is that intended?

xxalpha commented:
I'll be doing a PR in a week to deal with problems.
